File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.js
function solution(new_id) {
    let result = new_id
    .toLowerCase()
    .match(/[a-z 0-9 \- _ .]/g).join('')
    .replaceAll(/[.]+/g, ".")
    .replaceAll(/(^\.)|(\.$)/g, "")
    
    result = result.length === 0 ? "a" : result
    result = result.length >= 16 ?
        result.slice(0,15).replace(/\.$/g, "") : result
    while(result.length <= 2){
        result += result[result.length-1]
    }
    return result;
}

File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.py
def solution(new_id):
    
    # 1단계
    new_id = new_id.lower()
    
    # 2단계
    result = ""
    
    for i in new_id:
        if i.islower() or i.isdigit() or i in "-_.":
            result += i
            
    new_id = result
    
    # 3단계
    while ".." in new_id:
        new_id = new_id.replace("..", ".")
    
    # 4단계
    new_id = new_id.strip(".")
    
    # 5단계
    if new_id == "":
        new_id = "a"
    
    # 6단계
    new_id = new_id[:15]
    new_id = new_id.rstrip(".")
    
    # 7단계
    while len(new_id) < 3:
        new_id += new_id[-1]
    
    return new_id